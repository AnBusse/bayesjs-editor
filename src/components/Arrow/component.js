import React from 'react';
import { omit } from 'ramda';
import { noop } from 'lodash';

import PropTypes from 'prop-types';
import { getComponentTestId } from 'utils/test-utils';
import { positionPropTypes } from 'models';

const style = {
  transition: 'stroke-opacity 0.2s',
};
const stylePath = { cursor: 'pointer' };
const omitProps = omit(['markEnd', 'markEndStyle']);

const Arrow = ({
  title,
  markerEnd,
  pathD,
  onMouseOver,
  onMouseLeave,
  ...props
}) => (
  <g
    style={style}
    onMouseOver={onMouseOver}
    onFocus={onMouseOver}
    onMouseLeave={onMouseLeave}
    onBlur={onMouseLeave}
    data-testid={getComponentTestId('Arrow')}
    {...omitProps(props)}
  >
    <path
      d={pathD}
      fill="none"
      stroke="#333"
      strokeWidth="2"
      markerEnd={markerEnd}
      style={stylePath}
    >
      {title}
    </path>
  </g>
);

Arrow.defaultProps = {
  markEndStyle: '',
  onMouseOver: noop,
  onMouseLeave: noop,
  strokeOpacity: null,
};

Arrow.propTypes = {
  from: PropTypes.objectOf(positionPropTypes).isRequired,
  to: PropTypes.objectOf(positionPropTypes).isRequired,
  onMouseDown: PropTypes.func.isRequired,
  markEnd: PropTypes.bool.isRequired,
  markEndStyle: PropTypes.string,
  markerEnd: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  onMouseOver: PropTypes.func,
  onMouseLeave: PropTypes.func,
  pathD: PropTypes.string.isRequired,
  strokeOpacity: PropTypes.string,
};

export default Arrow;
