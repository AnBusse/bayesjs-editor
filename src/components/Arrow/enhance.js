import {
  withProps, pure, compose, withHandlers,
} from 'recompose';

import { makeArrowLine } from 'utils/arrows-positions';

const enhance = compose(
  pure,
  withProps(({
    from,
    to,
    markEnd,
    markEndStyle = 'url(#triangle)',
  }) => ({
    pathD: makeArrowLine(from, to),
    markerEnd: markEnd ? markEndStyle : '',
  })),
  withHandlers({
    onMouseOver: ({ onMouseOver, id }) => () => onMouseOver(id),
    onMouseLeave: ({ onMouseLeave, id }) => () => onMouseLeave(id),
  }),
);

export default enhance;
