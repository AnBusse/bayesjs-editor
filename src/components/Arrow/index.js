import component from './component';
import enhance from './enhance';

export default enhance(component);
