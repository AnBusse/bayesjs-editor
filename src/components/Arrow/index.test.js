import { shallow } from 'enzyme';
import React from 'react';
import Arrow from './component';

const defaultProps = {
  id: 'id',
  title: 'title',
  markerEnd: '',
  markEnd: false,
  from: { x: 0, y: 0 },
  to: { x: 100, y: 100 },
  onMouseDown: jest.fn(),
  pathD: '',
};

const shallowComponent = (props = {}) => {
  const compProps = { ...defaultProps, ...props };

  return shallow(<Arrow {...compProps} />);
};

describe('Arrow Component', () => {
  it('matches snapshot', () => {
    const component = shallowComponent();

    expect(component).toMatchSnapshot();
  });
});
