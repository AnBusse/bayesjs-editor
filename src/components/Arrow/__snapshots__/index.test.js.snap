// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrow Component matches snapshot 1`] = `
<g
  data-testid="arrow"
  id="id"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "transition": "stroke-opacity 0.2s",
    }
  }
>
  <path
    fill="none"
    markerEnd={true}
    stroke="#333"
    strokeWidth="2"
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    title
  </path>
</g>
`;
